### PR TITLE
Paste note fixes

### PR DESCRIPTION
**Pull request by wizofaus**
Resolves: https://github.com/musescore/MuseScore/issues/8824

Ensure that pasting notes works as expected in all circumstances

- [x] I signed [CLA](https://musescore.org/en/cla)
- [x] I made sure the code in the PR follows [the coding rules](https://github.com/musescore/MuseScore/wiki/CodeGuidelines)
- [x] I made sure the code compiles on my machine
- [x] I made sure there are no unnecessary changes in the code
- [x] I made sure the title of the PR reflects the core meaning of the issue you are solving
- [x] I made sure the commit message(s) contain a description and answer the question "Why do those changes fix that particular issue?" or "Why are those changes really necessary as improvements?"
- [x] I made sure the commit message title starts with "fix #424242:" if there is a related issue
- [x] I created the test (mtest, vtest, script test) to verify the changes I made

### CODE REVIEW

File: src/engraving/libmscore/chordrest.cpp [hunk 1]
```diff
@@ -1433,7 +1433,7 @@ void ChordRest::removeMarkings(bool /* keepTremolo */)
 bool ChordRest::isBefore(const ChordRest* o) const
 {
     if (!o || this == o) {
-        return true;
+        return false;
```
Comment (wizofaus) [inline src/engraving/libmscore/chordrest.cpp:1436]: Prevent asserts due to unstable std::sort predicates.

File: src/engraving/libmscore/cmd.cpp [hunk 2]
```diff
@@ -1049,6 +1053,13 @@ Fraction Score::makeGap(Segment* segment, int track, const Fraction& _sd, Tuplet
             tuplet = 0;
         } else {
             if (seg != firstSegment || !keepChord) {
+                if (cr->isChord()) {
+                    for (auto note : toChord(cr)->notes()) {
+                        _selection.remove(note);
+                    }
+                } else {
+                    _selection.remove(cr);
+                }
```
Comment (wizofaus) [inline src/engraving/libmscore/cmd.cpp:1062]: This logic should probably be in Score::removeElement( ) but wanted to minimize impact

File: src/engraving/libmscore/cmd.cpp [hunk 3]
```diff
@@ -3773,7 +3784,7 @@ void Score::cmdPitchUp()
     if (el && el->isLyrics()) {
         cmdMoveLyrics(toLyrics(el), Direction::UP);
     } else if (el && (el->isArticulation() || el->isTextBase())) {
-        el->undoChangeProperty(Pid::OFFSET, el->offset() + PointF(0.0, -MScore::nudgeStep* el->spatium()), PropertyFlags::UNSTYLED);
+        el->undoChangeProperty(Pid::OFFSET, el->offset() + PointF(0.0, -MScore::nudgeStep * el->spatium()), PropertyFlags::UNSTYLED);
```
Comment (wizofaus) [inline src/engraving/libmscore/cmd.cpp:3787]: hmm, uncrustify version incompatibilities strike again

File: src/engraving/libmscore/edit.cpp [hunk 4]
```diff
@@ -453,7 +453,7 @@ Rest* Score::setRest(const Fraction& _tick, int track, const Fraction& _l, bool
             // compute list of durations which will fit l
             //
             std::vector<TDuration> dList;
-            if (tuplet || staff->isLocalTimeSignature(tick)) {
+            if (tuplet || staff->isLocalTimeSignature(tick) || f == Fraction(0, 1)) {
```
Comment (wizofaus) [inline src/engraving/libmscore/edit.cpp:456]: was crashing without this when pasting, e.g. a whole note onto a multi-measure rest, the 'else' part of this test definitely doesn't work if `f` is 0 anyway

File: src/engraving/libmscore/paste.cpp [hunk 5]
```diff
@@ -1059,48 +1071,37 @@ void Score::cmdPaste(const QMimeData* ms, MuseScoreView* view, Fraction scale)
...
-                    // (ChordRest::drop() will select it for us)
-                    if (targetType == ElementType::NOTE) {
-                        select(nel);
+                    // duration information needs to be provided separately in this case
+                    if (target->isNote() && toNote(target)->chord()->ticks() != duration) {
+                        newEl = replaceNote(toNote(target)->chord(), toNote(newEl), duration);
+                    } else if (target->isChordRest()) {
+                        newEl = replaceNote(toChordRest(target), toNote(newEl), duration);
+                    } else {
+                        target->drop(ddata);
```
Comment (wizofaus) [inline src/engraving/libmscore/paste.cpp:1092]: would prefer that above logic was all inside Note::drop( ) and ChordRest::drop( ) but would require adding even more info (duration) to the already-bloated EditData structure...

**Comment by Jojo-Schmitz** [comment]
There's a merge commit that doesn't belong into this PR.
Is this supposed to fix https://musescore.org/en/node/321809 for master? Alternative to the 2nd commit for #8818?

**Comment by wizofaus** [comment]
> There's a merge commit that doesn't belong into this PR.
> Is this supposed to fix https://musescore.org/en/node/321809 for master? Alternative to the 2nd commit for #8818?

That came from a rebase.  I tried to squash but it wouldn't play nicely. Need to fix the code-style thing though.

**Comment by wizofaus** [comment]
And when I say "fix", I mean screw up my sensible formatting to suit the current version of uncrustify used by the git workflows.

**Comment by Jojo-Schmitz** [comment]
As far as I can tell this PR, on dropping a notehead of a quarter note onto an MMRest, pastes a quarter note, not a note equal to a the full lenthg of the measure?

**Comment by wizofaus** [comment]
It preserves the length of the note you copied, which is consistent with how it works when pasted onto anything else.
That's a new behaviour of MU4 (MU3 only ever transferred the pitch as the clipboard data didn't even include the note duration).

**Comment by Jojo-Schmitz** [comment]
But that then is not what you describe as the desired outcome in section a) of 8824

**Comment by wizofaus** [comment]
Yes it is, because pasting on to a whole measure rest in MU4 *does* preserve the length of the copied note (adding/removing rests as required to fit it on).

**Comment by Jojo-Schmitz** [comment]
_Pasting a note on to a multi-measure rest should behave the same as pasting onto a whole rest,_
That's not what it does

**Comment by Jojo-Schmitz** [comment]
Hmm, OK, not what it did in 3.x...

**Comment by wizofaus** [comment]
BTW this is not ready for merge yet, still a number of issues I just discovered.

**Comment by RobFog** [comment]
> BTW this is not ready for merge yet, still a number of issues I just discovered.

You can mark the pull request as a draft.

**Comment by wizofaus** [comment]
I'd looked for that option (convert to draft), couldn't see it, it's not very obvious!

File: src/engraving/libmscore/chordrest.cpp [hunk 6]
```diff
@@ -519,16 +519,17 @@ Element* ChordRest::drop(EditData& data)
...
-        delete e;
+        Segment* seg = segment();
+        score()->undoRemoveElement(this);
+        Chord* chord = new Chord(score());
+        chord->setTrack(track());
+        chord->setDurationType(durationType());
+        chord->setTicks(ticks());
+        chord->setTuplet(tuplet());
+        chord->add(note);
+        score()->undoAddCR(chord, seg->measure(), seg->tick());
```
Comment (wizofaus) [inline src/engraving/libmscore/chordrest.cpp:530]: existing method failed to transfer all properties and didn't work with tuplets etc.

File: src/engraving/libmscore/cmd.cpp [hunk 7]
```diff
@@ -777,15 +777,19 @@ Segment* Score::setNoteRest(Segment* segment, int track, NoteVal nval, Fraction
...
             expandVoice(segment, track);
         }
-
+        if (targetIsRest && !cr->isRest()) {
+            undoRemoveElement(cr);
+            segment = addRest(segment, track, cr->ticks(), cr->tuplet())->segment();
+        }
         // the returned gap ends at the measure boundary or at tuplet end
-        Fraction dd = makeGap(segment, track, sd, cr ? cr->tuplet() : 0);
+        Fraction dd = makeGap(segment, track, sd, tuplet);
```
Comment (wizofaus) [inline src/engraving/libmscore/cmd.cpp:792]: the changes here are about ensuring that a) if the element at the segment position is a rest, then any other notes that would be overlapped get removed instead of being "split" into component durations - actually I'd rather simplify this so that setNoteRest *always* does that but it may have unexpected side effects and b) only the *initial* tuplet is preserved where possible - the existing logic tried to preserve all tuplets that fell inside the duration, which didn't make sense for any case I could think of

File: src/engraving/libmscore/cmd.cpp [hunk 8]
```diff
@@ -1026,7 +1033,7 @@ Fraction Score::makeGap(Segment* segment, int track, const Fraction& _sd, Tuplet
             }
         }
         Tuplet* ltuplet = cr->tuplet();
-        if (ltuplet != tuplet) {
+        if (ltuplet != tuplet || (tuplet && sd > tuplet->ticks())) {
```
Comment (wizofaus) [inline src/engraving/libmscore/cmd.cpp:1036]: breaking change, investigating

File: src/engraving/libmscore/cmd.cpp.uncrustify [hunk 9]
```diff
@@ -0,0 +1,4438 @@
+/*
+ * SPDX-License-Identifier: GPL-3.0-only
```
Comment (Jojo-Schmitz) [inline src/engraving/libmscore/cmd.cpp.uncrustify:2]: I guess this file doesn't belong into the PR

**Comment by Jojo-Schmitz** [comment]
Issue: select note(head), use "Paste double/half duration". It keeps the duration rather than doubling/halfing

**Comment by wizofaus** [comment]
Good point, yes, I'd expect that to work.

File: src/engraving/libmscore/cmd.cpp.uncrustify (hunk 9 above)
Comment (wizofaus) [reply src/engraving/libmscore/cmd.cpp.uncrustify]: I'd removed it within minutes!

File: src/engraving/tests/tst_copypaste.cpp [hunk 10]
```diff
@@ -82,12 +83,21 @@ private slots:
...
+    void copypasteTuplet02() { copypastetuplet("02"); }
+    void copypasteQtrNoteOntoWholeRest() { copypastenote("01"); }
+    void copypasteQtrNoteOntoWholeNote() { copypastenote("02"); }
+    void copypasteQtrNoteOntoQtrRest() { copypastenote("03"); }
+    void copypasteQtrNoteOntoQtrNote() { copypastenote("04"); }
+    void copypasteWholeNoteOntoQtrNote() { copypastenote("05"); }
+    void copypasteWholeNoteOntoQtrRest() { copypastenote("06"); }
+    void copypasteQtrNoteOntoTriplet() { copypastenote("07"); }
+    void copypasteWholeNoteOntoTriplet() { copypastenote("08"); }
+    void copypasteQtrNoteIntoChord() { copypastenote("09"); }
```
Comment (wizofaus) [inline src/engraving/tests/tst_copypaste.cpp:98]: pretty minimal set of tests, there's quite a few more edge cases I'd like to have tests for, but for some them the "desired" behaviour isn't necessarily what we're getting now, I guess at least having a test to ensure it doesn't crash would be good!

File: src/engraving/libmscore/paste.cpp [hunk 11]
```diff
@@ -1031,6 +1031,60 @@ void Score::pasteSymbols(XmlReader& e, ChordRest* dst)
...
+    if (target->isNote() && toNote(target)->chord()->ticks() != duration) {
+        return prepareTarget(toNote(target)->chord(), with, duration);
+    }
+    if (target->isChordRest() && toChordRest(target)->ticks() != duration) {
+        return prepareTarget(toChordRest(target), with, duration);
+    }
+    return target;
+}
+
+static bool canPasteStaff(XmlReader& reader, const Fraction& scale)
```
Comment (wizofaus) [inline src/engraving/libmscore/paste.cpp:1069]: to prevent crashing when using "paste half length" on a range with 1/1024th notes, a bit kludgy but I tried a few other ways and this was the more reliable and least code by far

File: src/engraving/libmscore/paste.cpp [hunk 12]
```diff
@@ -1129,9 +1164,12 @@ void Score::cmdPaste(const QMimeData* ms, MuseScoreView* view, Fraction scale)
                 qDebug("paste <%s>", data.data());
             }
             XmlReader e(data);
-            e.setPasteMode(true);
-            if (!pasteStaff(e, cr->segment(), cr->staffIdx(), scale)) {
-                return;
+            if (canPasteStaff(e, scale)) {
+                XmlReader e(data);
```
Comment (wizofaus) [inline src/engraving/libmscore/paste.cpp:1168]: XmlReader doesn't have a rewind...should add one perhaps?

Comment (RomanPudashkin) [reply src/engraving/libmscore/paste.cpp:1168]: You already declared XmlReader e(data) at line 1166. We will get a warning here

File: src/engraving/libmscore/cmd.cpp.uncrustify (hunk 9 above)
Comment (Jojo-Schmitz) [reply src/engraving/libmscore/cmd.cpp.uncrustify]: But not pushed it equally fast I guess? Anyway, issue is sorted now.

File: src/engraving/libmscore/paste.cpp (hunk 12 above)
Comment (Jojo-Schmitz) [reply src/engraving/libmscore/paste.cpp:1168]: Declared? Rather called. So it might get called twice now.

Comment (wizofaus) [reply src/engraving/libmscore/paste.cpp:1168]: Correct,  you can't rewind it so you have to create a new one... happy to add a rewind function

Comment (wizofaus) [reply src/engraving/libmscore/paste.cpp:1168]: They're in different scopes so the behavior is well defined, but I agree stylistically some may prefer avoiding reusing the name (actually I don't like single letter variable names generally!).

Comment (wizofaus) [reply src/engraving/libmscore/paste.cpp:1173]: Took your suggestion of creating a separate method, creating a rewind method was a bit painful/dangerous due to inner workings of Qt.

**Review by RomanPudashkin — APPROVED** [review]

**Comment by vshalkevich** [comment]
Works fine
